File: src/jupyter_matlab_kernel/kernels/labextension_comm/actions/check_file_exists_action.py
# Copyright 2025 The MathWorks, Inc.

from . import ActionCommand, ActionTypes
from pathlib import Path


class CheckFileExistsAction(ActionCommand):
    def __init__(self, kernel):
        self.kernel = kernel
        self.log = kernel.log

    def get_code(self):
        pass

    async def execute(self, comm, data):
        """Starts MATLAB proxy
        Args:
            comm (ipykernel.comm.Comm): IPYKernels' Commuincation object
        """
        file_to_check = data["data"].get("mlxFilePath", "")

        if not file_to_check:
            error = Exception("No file path provided for checking if it exists")
            self.log.error(str(error))
            comm.send(
                {
                    "action": ActionTypes.CHECK_FILE_EXISTS.value,
                    "exists": False,
                    "error": str(error),
                }
            )

        else:
            comm.send(
                {
                    "action": ActionTypes.CHECK_FILE_EXISTS.value,
                    "exists": Path(file_to_check).expanduser().resolve().exists(),
                    "error": None,
                }
            )
            self.log.debug(f"File exists at {Path(file_to_check)}")

File: src/jupyter_matlab_labextension/src/plugins/actions/checkFileExistsAction.ts
// Copyright 2026 The MathWorks, Inc.

import { PromiseDelegate, ReadonlyJSONValue } from '@lumino/coreutils';

import { ICommunicationChannel } from '../matlabCommunication';
import { BaseAction } from './baseAction';
import { ActionTypes } from './actionTypes';
import { isCommValid } from './actionUtils';

export class CheckFileExistsAction extends BaseAction {
    blocking: boolean;
    private static blockingPromise: PromiseDelegate<ReadonlyJSONValue> | null;
    private static fileExists: boolean = false;

    constructor (blocking: boolean) {
        super();
        this.blocking = blocking;
    }

    public getActionName (): string {
        return ActionTypes.CHECK_FILE_EXISTS;
    }

    public static getFileExistsStatus (): boolean {
        return CheckFileExistsAction.fileExists;
    }

    public async execute (data: any, comm: ICommunicationChannel): Promise<void> {
        if (!('ipynbFilePath' in data)) {
            console.error('Missing data for executing CheckFileExists action...');
            return;
        }

        if (!CheckFileExistsAction.blockingPromise) {
            CheckFileExistsAction.blockingPromise =
        new PromiseDelegate<ReadonlyJSONValue>();
        }

        if (this.sendCheckFileExistsRequest(data.ipynbFilePath, comm)) {
            console.debug('Successfully sent check file exists request to kernel.');
        } else {
            console.error('Failed to send check file exists request to kernel.');
        }

        if (this.blocking) {
            console.debug(
                'CheckFileExists blocking action will wait for promise to resolve...'
            );
            await CheckFileExistsAction.blockingPromise.promise;
        }

        console.debug('CheckFileExistsAction execute completed');
    }

    public onMsg (data: any, _: ICommunicationChannel): void {
        if ('error' in data && data.error) {
            console.error('Received error from kernel ', data.error);
            if (CheckFileExistsAction.blockingPromise) {
                CheckFileExistsAction.blockingPromise.reject({ reason: data.error });
                CheckFileExistsAction.blockingPromise = null;
            }
            return;
        }
        CheckFileExistsAction.fileExists = data.exists;

        if (CheckFileExistsAction.blockingPromise) {
            CheckFileExistsAction.blockingPromise.resolve(null);
            CheckFileExistsAction.blockingPromise = null;
        }

        console.debug('CheckFileExistsAction onMsg completed');
    }

    private sendCheckFileExistsRequest (
        filePath: string,
        comm: ICommunicationChannel
    ): boolean {
        if (!isCommValid(comm)) {
            console.error('Communication channel is not available');
            return false;
        }

        if (!filePath) {
            console.error('File path is not available');
            return false;
        }

        comm.send({
            action: ActionTypes.CHECK_FILE_EXISTS,
            data: {
                ipynbFilePath: filePath
            }
        });
        return true;
    }
}
